File: problems/easy/plus_one.py
class Solution:
    def plusOne(self, digits: list[int]) -> list[int]:
        for i in range(len(digits)-1, -1, -1):
            if digits[i] == 9:
                digits[i] = 0
            else:
                digits[i] = digits[i] + 1
                return digits
        return [1] + digits
    

n1 = [1, 2, 3]
n2 = [9]
n3 = [9, 9]

sol = Solution()
assert sol.plusOne(n1) == [1, 2, 4]
assert sol.plusOne(n2) == [1, 0]
assert sol.plusOne(n3) == [1, 0, 0]

File: problems/easy/plus_one.go
package main

import "fmt"

func main() {
	var arr [1]int
	arr[0] = 9
	fmt.Println(plusOne(arr[:]))
}

func addFirst(s []int, insertValue int) []int {
	res := make([]int, len(s)+1)
	copy(res[1:], s)
	res[0] = insertValue
	return res
}

func plusOne(digits []int) []int {
	for i := len(digits) - 1; i >= 0; i-- {
		if digits[i] == 9 {
			digits[i] = 0
		} else {
			digits[i] = digits[i] + 1
			return digits
		}
	}
	return addFirst(digits, 1)
}